### PR TITLE
[IMP] api: New X2Many command constant

### PR DESCRIPTION
**Pull request by Julien00859**
One2many and Many2many use a special "commands" format to manipulate
the set of records stored in/linked to the field. There is a total of 7
commands: `create`, `update`, `delete`, `unlink`, `link`, and `set`.

* The `create` command (id: 0) is used to create a new record in the
  related model and to link it in the current record.
* The `update` command (id: 1) is used to update the targeted record id
  with new values.
* The `delete` command (id: 2) is used to both remove the target record
  id from the related table and to remove its link from the current
  record.
* The `unlink` command (id: 3) is used to remove the link to the
  targeted record id from the current record.
* The `link` command (id: 4) is used to link the targeted record id
  which exists already in the related model in the current record.
* The `clear` command (id: 5) is used to empty the list of linked record
  ids.
* The `set` command (id: 6) is used to replace the list of linked record
  ids with the provided.

Each command is a tuple with up to 3 elements, the second and/or third
element is ignored when it is not applicable to the command:

1) the mandatory command identifier
2) the targeted record id (1, 2, 3, 4)
3) create/write values dictionary (0, 1) or target record ids (6)

--

Before this contribution, the several commands did not have a
standardized name, the tuples were just hard-coded with the integer
identifier. For long developers have been complaining the commands were
hard to write and hard to read mainly due to the meaningless integer.

This contribution provide a new namespace with the constants and
functions desired to ease writting and reading the X2Many commands. All
modules have been updated to use the new constants instead of the
integer identifier.

On the frontend side, such constants (in the form of strings) do exist
already but do not conform to the new standardized names. It has been
decided to enforce the new nomenclature but to provide aliases to the
old names to ensure a better compatibility.

Task: 2366606

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -2410,7 +2410,7 @@ def _post(self, soft=True):
                     'line_ids': []
                 }
                 for line in move.line_ids.filtered(lambda line: line.account_id.user_type_id.type in ('receivable', 'payable')):
-                    to_write['line_ids'].append((1, line.id, {'name': to_write['payment_reference']}))
+                    to_write['line_ids'].append((api.CREATE, line.id, {'name': to_write['payment_reference']}))
```
Comment (Julien00859) [inline addons/account/models/account_move.py:2413]: api.UPDATE

File: addons/test_mass_mailing/tests/test_mailing_internals.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 from odoo.addons.test_mass_mailing.tests.common import TestMassMailCommon
 from odoo.addons.test_mass_mailing.data.mail_test_data import MAIL_TEMPLATE
 from odoo.tests.common import users
-from odoo.tools import mute_logger
```
Comment (Julien00859) [inline addons/test_mass_mailing/tests/test_mailing_internals.py:7]: wtf

File: addons/website_forum/models/forum.py [hunk 3]
```diff
@@ -349,7 +349,7 @@ def _search_can_view(self, operator, value):
         user = self.env.user
         # Won't impact sitemap, search() in converter is forced as public user
         if self.env.is_admin():
-            return [(1, '=', 1)]
+            return [(fields.X2ManyCmd.UPDATE, '=', 1)]
```
Comment (Julien00859) [inline addons/website_forum/models/forum.py:352]: wrong

File: odoo/__init__.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 ]
 
 import sys
-assert sys.version_info > (3, 6), "Outdated python version detected, Odoo requires Python >= 3.6 to run."
+assert sys.version_info > (fields.X2ManyCmd.UNLINK, 6), "Outdated python version detected, Odoo requires Python >= 3.6 to run."
```
Comment (Julien00859) [inline odoo/__init__.py:19]: wrong

File: odoo/addons/base/models/ir_fields.py [hunk 5]
```diff
@@ -17,12 +17,12 @@ def exclude_ref_fields(record):
...
-FORGET = lambda id: (3, id, False)
-LINK_TO = lambda id: (4, id, False)
-DELETE_ALL = lambda: (5, False, False)
-REPLACE_WITH = lambda ids: (6, False, ids)
+UPDATE = lambda id, values: (fields.X2ManyCmd.UPDATE, id, values)
+DELETE = lambda id: (fields.X2ManyCmd.DELETE, id, False)
+FORGET = lambda id: (fields.X2ManyCmd.UNLINK, id, False)
+LINK_TO = lambda id: (fields.X2ManyCmd.LINK, id, False)
+DELETE_ALL = lambda: (fields.X2ManyCmd.CLEAR, False, False)
+REPLACE_WITH = lambda ids: (fields.X2ManyCmd.SET, False, ids)
```
Comment (Julien00859) [inline odoo/addons/base/models/ir_fields.py:25]: should rewrite the file to use the command function defined on `fields.X2ManyCmd`

File: odoo/addons/base/tests/test_api.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models
+from odoo import api, models
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_api.py:4]: missing fields, useless api

File: odoo/addons/base/tests/test_cloc.py [hunk 7]
```diff
@@ -85,9 +85,9 @@ def test_parser(self):
         xml_count = cl.parse_xml(XML_TEST)
         self.assertEqual(xml_count, (18, 31))
         py_count = cl.parse_py(PY_TEST_NO_RETURN)
-        self.assertEqual(py_count, (2, 2))
+        self.assertEqual(py_count, (fields.X2ManyCmd.DELETE, 2))
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_cloc.py:88]: wrong

File: odoo/addons/base/tests/test_cloc.py [hunk 8]
```diff
@@ -85,9 +85,9 @@ def test_parser(self):
...
         py_count = cl.parse_py(PY_TEST)
-        if sys.version_info >= (3, 8, 0):
+        if sys.version_info >= (fields.X2ManyCmd.UNLINK, 8, 0):
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_cloc.py:90]: wrong

File: odoo/addons/base/tests/test_expression.py [hunk 9]
```diff
@@ -579,15 +579,15 @@ def test_20_expression_parse(self):
...
+
         # Special =? operator mean "is equal if right is set, otherwise always True"
         users = self._search(Users, [('name', 'like', 'test'), ('parent_id', '=?', False)])
         self.assertEqual(users, a + b1 + b2, '(x =? False) failed')
         users = self._search(Users, [('name', 'like', 'test'), ('parent_id', '=?', b1.partner_id.id)])
         self.assertEqual(users, b2, '(x =? id) failed')
 
     def test_30_normalize_domain(self):
-        norm_domain = domain = ['&', (1, '=', 1), ('a', '=', 'b')]
+        norm_domain = domain = ['&', (fields.X2ManyCmd.UPDATE, '=', 1), ('a', '=', 'b')]
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_expression.py:590]: wrong

File: odoo/addons/base/tests/test_expression.py [hunk 10]
```diff
@@ -928,14 +928,14 @@ def patch_domain(model, fname, value):
         patch_domain(partner_obj, 'bank_ids', [('sanitized_acc_number', 'like', '2')])
 
         # Do: 2 cascaded one2many with _auto_join, test final leaf is an id
-        partners = partner_obj.search(['&', (1, '=', 1), ('child_ids.bank_ids.id', 'in', [b_aa.id, b_ba.id])])
+        partners = partner_obj.search(['&', (fields.X2ManyCmd.UPDATE, '=', 1), ('child_ids.bank_ids.id', 'in', [b_aa.id, b_ba.id])])
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_expression.py:931]: wrong

File: odoo/addons/base/tests/test_expression.py [hunk 11]
```diff
@@ -928,14 +928,14 @@ def patch_domain(model, fname, value):
...
         self.assertLessEqual(p_a, partners,
             "_auto_join on one2many with domains incorrect result")
         self.assertFalse((p_ab + p_ba) & partners,
             "_auto_join on one2many with domains incorrect result")
 
         patch_domain(partner_obj, 'child_ids', lambda self: [('name', '=', '__%s' % self._name)])
-        partners = partner_obj.search(['&', (1, '=', 1), ('child_ids.bank_ids.id', 'in', [b_aa.id, b_ba.id])])
+        partners = partner_obj.search(['&', (fields.X2ManyCmd.UPDATE, '=', 1), ('child_ids.bank_ids.id', 'in', [b_aa.id, b_ba.id])])
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_expression.py:938]: wrong

File: odoo/addons/base/tests/test_image.py [hunk 12]
```diff
@@ -72,7 +72,7 @@ def test_02_image_fix_orientation(self):
         """Test that the orientation of images is correct."""
 
         # Colors that can be distinguished among themselves even with jpeg loss.
-        blue = (0, 0, 255)
+        blue = (fields.X2ManyCmd.CREATE, 0, 255)
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_image.py:75]: wrong

File: odoo/addons/base/tests/test_image.py [hunk 13]
```diff
@@ -234,7 +234,7 @@ def test_15_image_process_colorize(self):
         # CASE: color random, color has changed
         image = tools.base64_to_image(tools.image_process(base64_rgba, colorize=True))
         self.assertEqual(image.mode, 'RGB')
-        self.assertNotEqual(image.getpixel((0, 0)), (0, 0, 0))
+        self.assertNotEqual(image.getpixel((fields.X2ManyCmd.CREATE, 0)), (fields.X2ManyCmd.CREATE, 0, 0))
```
Comment (Julien00859) [inline odoo/addons/base/tests/test_image.py:237]: wrong

File: odoo/addons/test_impex/tests/test_load.py [hunk 14]
```diff
@@ -1105,7 +1105,7 @@ def test_recursive_o2m(self):
         self.assertEqual(len(result['ids']), 2)
 
         b = Model.browse(result['ids'])
-        self.assertEqual((b[0].value, b[1].value), (4, 5))
+        self.assertEqual((b[0].value, b[1].value), (fields.X2ManyCmd.LINK, 5))
```
Comment (Julien00859) [inline odoo/addons/test_impex/tests/test_load.py:1108]: wrong

File: odoo/addons/test_lint/tests/test_pylint.py [hunk 15]
```diff
@@ -53,7 +53,7 @@ def test_pylint(self):
         if pylint is None:
             self._skip_test('please install pylint')
         required_pylint_version = LooseVersion('1.6.4')
-        if sys.version_info >= (3, 6):
+        if sys.version_info >= (fields.X2ManyCmd.UNLINK, 6):
```
Comment (Julien00859) [inline odoo/addons/test_lint/tests/test_pylint.py:56]: wrong

File: odoo/addons/test_new_api/models/test_new_api.py [hunk 16]
```diff
@@ -176,7 +176,7 @@ def _search_size(self, operator, value):
         self.env.cr.execute(query, (value,))
         ids = [t[0] for t in self.env.cr.fetchall()]
         # return domain with an implicit AND
-        return [('id', 'in', ids), (1, '=', 1)]
+        return [('id', 'in', ids), (fields.X2ManyCmd.UPDATE, '=', 1)]
```
Comment (Julien00859) [inline odoo/addons/test_new_api/models/test_new_api.py:179]: wrong

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 17]
```diff
@@ -11,7 +11,7 @@
 from PIL import Image
 import psycopg2
 
-from odoo import models, fields
+from odoo import api, models, fields
```
Comment (Julien00859) [inline odoo/addons/test_new_api/tests/test_new_fields.py:14]: useless

File: odoo/addons/test_new_api/tests/test_onchange.py [hunk 18]
```diff
@@ -217,7 +217,7 @@ def test_onchange_one2many_reference(self):
         result = self.Discussion.onchange(values, 'name', field_onchange)
         self.assertIn('messages', result['value'])
         self.assertItemsEqual(result['value']['messages'], [
-            (5,),
+            (fields.X2ManyCmd.CLEAR,),
             (0, REFERENCE, {
```
Comment (Julien00859) [inline odoo/addons/test_new_api/tests/test_onchange.py:221]: missing command create

File: odoo/fields.py [hunk 19]
```diff
@@ -2850,22 +2888,22 @@ def convert_to_cache(self, value, record, validate=True):
             # modify ids with the commands
             for command in value:
                 if isinstance(command, (tuple, list)):
-                    if command[0] == 0:
+                    if command[0] == fields.X2ManyCmd.CREATE:
```
Comment (Julien00859) [inline odoo/fields.py:2891]: (and bellow), useless `fields.`

File: odoo/fields.py [hunk 20]
```diff
@@ -2934,11 +2972,11 @@ def convert_to_write(self, value, record):
                             if name not in inv_names and record[name] != origin[name]
                         })
                         if values:
-                            result.append((1, origin.id, values))
+                            result.append((apt.UPDATE, origin.id, values))
```
Comment (Julien00859) [inline odoo/fields.py:2975]: apt ? wtf ?

File: odoo/osv/expression.py [hunk 21]
```diff
@@ -172,7 +172,7 @@
     'not ilike': 'ilike',
 }
 
-TRUE_LEAF = (1, '=', 1)
+TRUE_LEAF = (fields.X2ManyCmd.UPDATE, '=', 1)
```
Comment (Julien00859) [inline odoo/osv/expression.py:175]: Wrong

File: odoo/service/server.py [hunk 22]
```diff
@@ -459,7 +459,7 @@ def start(self, stop=False):
             signal.signal(signal.SIGUSR1, log_ormcache_stats)
         elif os.name == 'nt':
             import win32api
-            win32api.SetConsoleCtrlHandler(lambda sig: self.signal_handler(sig, None), 1)
+            win32fields.X2ManyCmd.SETConsoleCtrlHandler(lambda sig: self.signal_handler(sig, None), 1)
```
Comment (Julien00859) [inline odoo/service/server.py:462]: wtf ?!

File: odoo/tests/common.py [hunk 23]
```diff
@@ -149,7 +149,7 @@ def new_test_user(env, login='', groups='base.group_user', context=None, **kwarg
 # ------------------------------------------------------------
 class OdooSuite(unittest.suite.TestSuite):
 
-    if sys.version_info < (3, 8):
+    if sys.version_info < (fields.X2ManyCmd.UNLINK, 8):
```
Comment (Julien00859) [inline odoo/tests/common.py:152]: wrong

File: odoo/tests/common.py [hunk 24]
```diff
@@ -245,7 +245,7 @@ def _tearDownPreviousClass(self, test, result):
 
 class TreeCase(unittest.TestCase):
 
-    if sys.version_info < (3, 8):
+    if sys.version_info < (fields.X2ManyCmd.UNLINK, 8):
```
Comment (Julien00859) [inline odoo/tests/common.py:248]: wrong

File: odoo/tools/appdirs.py [hunk 25]
```diff
@@ -14,7 +14,7 @@
 # - Mac OS X: http://developer.apple.com/documentation/MacOSX/Conceptual/BPFileSystem/index.html
 # - XDG spec for Un*x: http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
 
-__version_info__ = (1, 3, 0)
+__version_info__ = (fields.X2ManyCmd.UPDATE, 3, 0)
```
Comment (Julien00859) [inline odoo/tools/appdirs.py:17]: wrong

File: odoo/tools/mail.py [hunk 26]
```diff
@@ -198,7 +198,7 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
         else:
             kwargs['remove_tags'] = tags_to_kill + tags_to_remove
 
-    if sanitize_attributes and etree.LXML_VERSION >= (3, 1, 0):  # lxml < 3.1.0 does not allow to specify safe_attrs. We keep all attributes in order to keep "style"
+    if sanitize_attributes and etree.LXML_VERSION >= (fields.X2ManyCmd.UNLINK, 1, 0):  # lxml < 3.1.0 does not allow to specify safe_attrs. We keep all attributes in order to keep "style"
```
Comment (Julien00859) [inline odoo/tools/mail.py:201]: wrong

**Comment by Julien00859** [comment]
To the attention of the reviewer, I am going to fix the conflict once for all when we agree on the PR and that I have a delegate+. The file of interest is `fields.py`, the rest is just match/replace, I did manually verify every single file.

**Comment by Yenthe666** [comment]
@Julien00859 do I get it right that this introduced helpers to do CRUD operations on X2many fields (so one2many & many2many)?

**Comment by Julien00859** [comment]
Hello Yenthe, yes you are, the file of interest is `fields.py`:

    class X2ManyCmd:
        CREATE = 0
        UPDATE = 1
        DELETE = 2
        UNLINK = 3
        LINK = 4
        CLEAR = 5
        SET = 6

        @staticmethod
        def create(values):
            return (0, 0, values)

        @staticmethod
        def update(id, values):
            return (1, id, values)

        @staticmethod
        def delete(id):
            return (2, id)

        @staticmethod
        def unlink(id):
            return (3, id)

        @staticmethod
        def link(id):
            return (4, id)

        @staticmethod
        def clear():
            return (5,)

        @staticmethod
        def set(ids):
            return (6, 0, ids)

File: odoo/fields.py [hunk 27]
```diff
@@ -2915,7 +2953,7 @@ def convert_to_write(self, value, record):
         if isinstance(value, BaseModel) and value._name == self.comodel_name:
             # make result with new and existing records
             inv_names = {field.name for field in record._field_inverses[self]}
-            result = [(6, 0, [])]
+            result = [(X2ManyCmd.SET, 0, [])]
```
Comment (rco-odoo) [inline odoo/fields.py:2956]: We should use the command functions here:
```suggestion
            ids = []
            result = [Command.set(ids)]
```
and use `ids.append()` below

File: odoo/fields.py [hunk 28]
```diff
@@ -2924,7 +2962,7 @@ def convert_to_write(self, value, record):
                         for name in record._cache
                         if name not in inv_names
                     })
-                    result.append((0, 0, values))
+                    result.append((X2ManyCmd.CREATE, 0, values))
```
Comment (rco-odoo) [inline odoo/fields.py:2965]:
```suggestion
                    result.append(Command.create(values))
```
and so on

**Review by rco-odoo — COMMENTED** [review]
We should use `Command.create()`, `Command.update()`, etc. instead of the tuples.
The constants `Command.CREATE`, `Command.UPDATE`, etc. should only be used where we parse those commands.

**Comment by Yenthe666** [comment]
Love your idea @rco-odoo, cleaner :)

File: odoo/fields.py (hunk 27 above)
Comment (Julien00859) [reply odoo/fields.py:2956]: Bruh... I don't like that kind of magic where a mutable structure is sent empty to a function and later is feeded with data.

File: odoo/fields.py (hunk 28 above)
Comment (Julien00859) [reply odoo/fields.py:2965]: I agree, this is cleaner... but also more complicated for my script to rewrite :p

File: odoo/fields.py (hunk 27 above)
Comment (MiquelRForgeFlow) [reply odoo/fields.py:2956]: why not `result = [Command.set([])]`?

**Pull request closed without merging**